### DISCUSSION TITLE
Avoiding user-gem conflicts with system Ruby

### DISCUSSION THREAD

**Discussion by konsolebox** [rbenv/rbenv#1520] [General]
I'm thinking of ways so user gems get to be installed to a different directory and so rbenv's installed Rubies don't use the default `user_dir` which is either `"#{ENV['HOME']}/.gem/#{RbConfig::CONFIG['ruby_version']}"` or `"#{ENV['HOME']}/.local/share/gem/#{RbConfig::CONFIG['ruby_version']}"`.

One of it is to append "-rbenv" to the version of every rbenv-installed Ruby.  Question is, is it the best approach for rbenv, and is it (or ruby-build) currently configurable to do so?

If it's not configurable, are there parts of the code (in either rbenv or ruby-build) that I could examine so I can perhaps modify them?

For reference, here's the definition of `Gem.user_dir`.  I don't think overriding the way these values are generated is a good idea as these values might again change in the future.  Better just change `RbConfig::CONFIG["ruby_version"]`.

```
def self.user_dir
  gem_dir = File.join(Gem.user_home, ".gem")
  gem_dir = File.join(Gem.data_home, "gem") unless File.exist?(gem_dir)
  parts = [gem_dir, ruby_engine]
  parts << RbConfig::CONFIG["ruby_version"] unless RbConfig::CONFIG["ruby_version"].empty?
  File.join parts
end
```

Also here's the definition of `Gem.data_home`:

```
def self.data_home
  @data_home ||= (ENV["XDG_DATA_HOME"] || File.join(Gem.user_home, ".local", "share"))
end
```

**Comment by konsolebox** [comment]
> I don't think overriding the way these values are generated is a good idea as these values might again change in the future.

I changed my mind about this.  Changing is OK as long as the new generator simply appends a suffix to it.  I was thinking about `user_dir` testing if `~/.gem/*` exists but it turns out it only tests `~/.gem` so appending a string to the final path wouldn't matter on it.

In Gentoo I'm solving this issue by appending `"-system"` to the `user_dir` with the following `defaults/operating_system.rb`, but this solutions requires making the system Ruby adapt to user Rubies instead of Rbenv adapting to system Ruby.  I wouldn't be able to do this in other systems.

```
# frozen_string_literal: true

module Gem
  module OperatingSystemDefaults
    def user_dir
      gentoo_user_dir
    end

    def default_dir
      gentoo_user_dir
    end

    def default_bindir
      gentoo_user_bindir
    end

    def default_path
      path = []
      path << gentoo_user_dir unless gentoo_local_dir == gentoo_user_dir
      path << gentoo_local_dir
      path << gentoo_dir
      path << vendor_dir if vendor_dir && File.directory?(vendor_dir)
      path
    end

    def operating_system_defaults
      { "custom_shebang" => File.join(_rbconfig("bindir"), "ruby") }
    end

    def default_specifications_dir
      orig_default_specifications_dir
    end

  private
    def _rbconfig(key)
      RbConfig::CONFIG[key] or raise "RbConfig::CONFIG[#{key.inspect}] unset."
    end

    def gentoo_dir
      @gentoo_dir ||= _rbconfig("sitelibdir").gsub("site_ruby", "gems")
    end

    def gentoo_local_dir
      @gentoo_local_dir ||= begin
        prefix = _rbconfig("exec_prefix")

        if gentoo_dir[0, prefix.size] == prefix
          File.join(prefix, "local", gentoo_dir[prefix.size..-1])
        else
          orig_user_dir_with_suffix
        end
      end
    end

    def gentoo_user_dir
      @gentoo_user_dir ||= Process.euid.zero? ? gentoo_local_dir : orig_user_dir_with_suffix
    end

    def gentoo_user_bindir
      @gentoo_user_bindir ||= begin
        if Process.euid.zero? && gentoo_local_dir != orig_user_dir_with_suffix
          File.join(_rbconfig("exec_prefix"), "local", "bin")
        else
          File.join(user_home, "bin")
        end
      end
    end

    def orig_user_dir_with_suffix
      @orig_user_dir_with_suffix ||= method(:user_dir).super_method.call + "-system"
    end

    def orig_default_specifications_dir
      @orig_default_specifications_dir ||= File.join(_rbconfig("rubylibprefix"), "gems",
          _rbconfig("ruby_version"), "specifications", "default")
    end
  end

  self.singleton_class.prepend OperatingSystemDefaults
end
```

**Comment by mislav** [comment]
> I'm thinking of ways so user gems get to be installed to a different directory and so rbenv's installed Rubies don't use the default `user_dir`

Hi, thanks for exploring this option! I have no strong opinions on this yet, but I'm curious to know and understand more.

You mentioned a "conflict" of sorts. How does this conflict play out for a typical user? I'm wondering what is the concrete issue being solved here.

**Reply by konsolebox** [reply]
If I recall right what drove me to separating user gem dirs is when I was trying to install yarp (see https://github.com/ruby/yarp/discussions/1385). Since Gentoo's Ruby has `--disable-static enabled`, the gem fails to install properly and the gem fails at runtime, but rbenv or ruby-build doesn't have the option so it doesn't fail.  For a while it confused me why rbenv's version works and then it suddenly doesn't. It took me a while to realize that the system's version of yarp taints rbenv since it writes to the same user directory.  Debugging would have been much easier if they weren't using the same user gem directory.

**Reply by mislav** [reply]
Thanks for explaining! That's a very specific problem to have, and one that I'm not sure ruby-build should be solving out-of-the-box. However, you are welcome to make a rbenv plugin that publishes an `install` hook to create a suitable `defaults/operating_system.rb` for every new installed rbenv version. With such a plugin, you could rest sure that each Ruby version uses its own user-install directory for gems.

**Reply by konsolebox** [reply]
Thanks. I agree that using a plugin is the way.
